### ISSUE TITLE
Proxy can not properly pass the Header `Host` to origin server

### ISSUE THREAD

**Issue by yamada95** [elazarl/goproxy#132]
In my use case, I start the proxy on `127.0.0.1:8080` and issue the following command:
`curl --proxy "127.0.0.1:8080" -H "Host: domain.com" "http://130.130.10.10/ping"`

I expect that the origin server `130.130.10.10` would receive the request contain the Header `Host: domain.com`, but I observed the Header `Host` was erased by the proxy.

After some search, `http.Request` has a `Host` field(https://golang.org/pkg/net/http/#Request), but it didn't help. 

If `req.URL.Host` is not empty, the `req.Host` will be set to this value(so in the example above, the `req.Host` is set to `130.130.10.10`, but what I expect is `domain.com`), and the `Host` in `req.Header` will be deleted(https://golang.org/src/net/http/request.go?s=2057:8270#L703), the origin Header `Host` is not available anymore.

Through the document, it's true that the `Host` in Headers should be ignored on the origin server if the request URI contains the host(https://www.ietf.org/rfc/rfc2616.txt [section 5.2]), but not in case of the proxy.

Is there a good way to solve this?

**Comment by elazarl** [comment]
Let me understand what you want more clearly.

You're sending a proxy request to `http://foo.com`, with header `Host: bar.com`, and you want the proxy to contact `foo.com` giving it the host name `bar.com`?

In that case, I think you can use a proxy handler to achieve the required behavior.

Or am I missing something?

**Comment by serika00** [comment]
Yes, I want the proxy contact `foo.com` with host name `bar.com`. But the host name `bar.com` is set by the client, not the proxy, the proxy know nothing about the host name, just deliver it.

**Comment by elazarl** [comment]
Hmm... the Go stdlib is actively removing the information you need in this
case.

It already made almost impossible to support ICECAST exception.

I think a good solution wouid be to copy net/http to a local package, and
change the necessary code to make it more flexible.

What do you think?

**Comment by serika00** [comment]
Copying net/http should work. Still finding more elegant way to do this.

Maybe the origin header `Host` should not be deleted in http/request.go. It's better to put this logic in http/server.go, let the `Request` be a more general purpose struct to handle http request.

**Comment by elazarl** [comment]
But how can you change the http server behaviour?

**Comment by yeyuqiu** [comment]
> ICECAST

Do you solve it ? Can you share with me ?

**Comment by yeyuqiu** [comment]
> Copying net/http should work. Still finding more elegant way to do this.
> 
> Maybe the origin header `Host` should not be deleted in http/request.go. It's better to put this logic in http/server.go, let the `Request` be a more general purpose struct to handle http request.

Do you solve it ? Can you share with me ?
